library/src/main/java/com/novoda/downloadmanager/VersionOneToVersionTwoMigrator.java

```diff
@@ -0,0 +1,137 @@
+package com.novoda.downloadmanager;
+
+import android.util.Log;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.List;
+
+class VersionOneToVersionTwoMigrator implements Migrator {
+
+    private static final int BUFFER_SIZE = 8 * 512;
+
+    private final MigrationExtractor migrationExtractor;
+    private final RoomDownloadsPersistence downloadsPersistence;
+    private final InternalFilePersistence internalFilePersistence;
+    private final DatabaseWrapper database;
+    private final Callback migrationCompleteCallback;
+
+    VersionOneToVersionTwoMigrator(MigrationExtractor migrationExtractor,
+                                   RoomDownloadsPersistence downloadsPersistence,
+                                   InternalFilePersistence internalFilePersistence,
+                                   DatabaseWrapper database,
+                                   Callback migrationCompleteCallback) {
+        this.migrationExtractor = migrationExtractor;
+        this.downloadsPersistence = downloadsPersistence;
+        this.internalFilePersistence = internalFilePersistence;
+        this.database = database;
+        this.migrationCompleteCallback = migrationCompleteCallback;
+    }
+
+    @Override
+    public void migrate() {
+        List<Migration> migrations = migrationExtractor.extractMigrations();
+
+        migrateV1FilesToV2Location(migrations);
+        migrateV1DataToV2Database(migrations);
+
+        deleteFrom(database, migrations);
+        database.close();
+
+        migrationCompleteCallback.onMigrationComplete();
+    }
+
+    private void migrateV1FilesToV2Location(List<Migration> migrations) {
+        for (Migration migration : migrations) {
+            migrateV1FilesToV2Location(migration);
+        }
+    }
+
+    private void migrateV1FilesToV2Location(Migration migration) {
+        Batch batch = migration.batch();
+
+        for (int i = 0; i < migration.originalFileLocations().size(); i++) {
```

Comment:
You can `migration.originalFileLocations().size()` once

Reply:
ace556cc2d0b4d63f6356b21eb82459afdaa5d38